### ISSUE TITLE
Restrict behavior of callable classes - specification text

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#32153]
This is the specification component of meta-bug #32152.  See the meta-bug for details.

**Comment by eernstg** [comment]
For additional detail there is also #32151 (which is being closed now as a duplicate of this one).

**Comment by stereotype441** [comment]
Just ran across a corner case that I could use clarification on.  What should happen in the case where a class contains `.call`, but its `.call` is a getter rather than a method?  What should happen if this getter isn't function-typed?

Examples:
```dart
class A {
  void call() {}
}
class B {
  void Function() get call => () {};
}
class C {
  dynamic get call => () {};
}
class D {
  A get call => new A();
}
class E {
  dynamic get call => new A();
}
main() {
  A a = new A();
  B b = new B();
  C c = new C();
  D d = new D();
  E e = new E();
  a(); // ok - equivalent to a.call()
  b(); // equivalent to b.call() ?
  c(); // equivalent to c.call() ?
  d(); // equivalent to d.call.call() ?
  e(); // ???
  Function f1 = a; // ok - tears off a.call
  Function f2 = b; // tears off b.call ?
  Function f3 = c; // tears off c.call ?
  Function f4 = d; // tears off d.call.call ?
  Function f5 = e; // ???
}
```

**Comment by eernstg** [comment]
I'm just working on a [CL for dartLangSpec.tex about callable objects](https://dart-review.googlesource.com/c/sdk/+/41360), and the approach that I'm currently proposing is that a `call` method admits function expression invocation, and so does a `call` getter, but the latter must yield a function object in one step (so we are not going to try `f.call.call.call....`, neither dynamically nor statically).

**Comment by lrhn** [comment]
My opinion is that we should get rid of this getter nonsense.
I'd prefer that an object is callable iff it declares a `call` *method*, a `call` getter does not count.

A `call` getter typed as `dynamic` or `Object` or `Function` would not specify an actual function type, and it might return different values each time it's invoked. Even with the "must return function first" rule, we then have to explain why a getter returning a callable object doesn't work. It's just not worth its own complexity.

It also has the advantage that the tear-off operation cannot fail. If the static type of the object is a callable object type, then the tear-off is non-null and assignable to the expected function type. Also, we don't have to distinguish whether an object is callable by the *return type* of the `call` getter (what if a subclass specializes it to return `Null`? Is the subclass not callable?), just on whether it has a `call` method at all. It's simpler.

You can still write `dynamic o = ...; o.call(42);` to call the getter, but you won't get `o(42)` as shorthand.

**Comment by leafpetersen** [comment]
SGTM (what Lasse said).

**Comment by stereotype441** [comment]
SGTM as well--makes things much simpler.  I'll put together additional test cases and a fix for analyzer/front_end.

**Comment by eernstg** [comment]
Cool, let's do that. CL update later today.

**Comment by eernstg** [comment]
I updated the [callable objects CL](https://dart-review.googlesource.com/c/sdk/+/41360) such that getters named `call` are no longer used during function invocation.

**Comment by eernstg** [comment]
Closing: Said [callable objects CL](https://dart-review.googlesource.com/c/sdk/+/41360) has been landed as cec672c7934e4c8513a58b271908efba702246a9. It was updated once more, to clarify that it is not implied that a class which `implements Function` has a method named `call`.
